### DISCUSSION TITLE
Model Loading Issue with Model URIs on Windows

### DISCUSSION THREAD

**Discussion by joerg-holtmann** [eclipse-emfcloud/emfcloud#97] [Coffee Editor] [answered]
Hey all,

With the setup of the [coffee editor model server example](https://github.com/eclipse-emfcloud/coffee-editor), I can access easily all loaded models via <code>http://localhost:8081/api/v1/models</code>. However, when I try to load a specific model via, e.g., <code>http://localhost:8081/api/v1/models?modeluri=model/Coffee.ecore</code>, this works not on my Windows machine. In contrast, this works fine with the same setup on a colleague's Mac.

Debugging into <code>org.eclipse.emfcloud.modelserver.emf.common.ModelRepository#getModel(final String modeluri)</code>, I see that the <code>modelResourceManager</code> has a resource <code>file:<b>/</b>\<absolutePathTo\>/model/Coffee.ecore</code>, but the submitted URI in the parameter <code>modeluri</code> has <code>file:<b>///</b>\<absolutePathTo\>/model/Coffee.ecore</code>. This apparently leads to not finding the key in the corresponding HashMap. I also tried many other variations in different browsers regarding the slashes, but the resource is never found. The issue also leads to the problem that I cannot access the model via the [Theia model client](https://github.com/eclipse-emfcloud/emfcloud-modelserver-theia), which again works fine on my colleague's Mac.

Is this a known issue and reproducible for others (maybe related to something far more general than EMF.cloud, like Windows-style directory paths and/or URL encoding)? Does anybody maybe know a lightweight workaround (besides adapting or extending ModelRepository to fix the submitted URLs on Windows)?

Best, Jörg

**Comment by ndoschek** [comment, marked as answer]
Hi Jörg!

The coffee editor uses an older version of the Model Sever in the current master.
A [fix in the Model Server](https://github.com/eclipse-emfcloud/emfcloud-modelserver/commit/4d5824189afc714931fc64d9392647226e295a98) is available that tackles problems with the URI handling on Windows.
As a workaround, you could update the target platform (`backend/releng/org.eclipse.emfcloud.coffee.target/org.eclipse.emfcloud.coffee.target.(tpd|target)`) to the Model Server version `0.7.0.202102171625`. This will include the mentioned fix regarding URIs under Windows and should not cause you any other necessary changes due to the update.

Hope that helps in the meantme, and if issues still come up, do not hesitate to contact us again.

Best, Nina

**Reply by joerg-holtmann** [reply]
Hi Nina,

thank you very much for the hint! The workaround works partially, i.e., it now works for absolute URIs (which did not work beforehand) but not for relative ones (which work fine on a Mac). However, absolute URIs are not really the end solution... ;-) Or is it maybe intended to construct an absolute URI on the client side? But at least I can continue my experiments now.

URIs for reproducing the problem with their respective encoding in <code>org.eclipse.emfcloud.modelserver.emf.common.ModelRepository#getModel(final String modeluri)</code>  (we might turn this into an issue?):
- <code>file:/\<absolutePathTo\>/model/Coffee.ecore</code> -> resolves to <code>file:/\<absolutePathTo\>/model/Coffee.ecore</code> -> works fine
- <code>file:///\<absolutePathTo\>/model/Coffee.ecore</code> -> resolves to <code>file:/\<absolutePathTo\>/model/Coffee.ecore</code> -> works fine
- <code>model/Coffee.ecore</code> -> resolves to <code>file:<b>///</b>\<absolutePathTo\>/model/Coffee.ecore</code> -> fails on Windows, but works on a Mac

Best, Jörg

**Reply by ndoschek** [reply]
Hi Jörg!

Great that we could find a fix in the meantime for your experiments.
Yes, I agree with you, this is definitely an issue then as it works also properly on Linux systems. I opened an issue directly at the `@emfcloud-modelserver` (https://github.com/eclipse-emfcloud/emfcloud-modelserver/issues/122).

Have a good weekend! Best, Nina
